nomad/event_sink_manager.go

```diff
@@ -0,0 +1,565 @@
+	eventSinksWs memdb.WatchSet
+
+	// running specifies if the manager is running
+	running bool
+
+	// shutdownCh is used to stop the manager from running
+	shutdownCh chan struct{}
+
+	// delegate is the interface needed to interact with State and RPCs
+	delegate SinkDelegate
+
+	L hclog.Logger
+}
+
+// SinkDelegate is the interface needed for the SinkManger to interfact with
+// parts of Nomad
+type SinkDelegate interface {
+	State() *state.StateStore
+	getLeaderAcl() string
+	Region() string
+	RPC(method string, args interface{}, reply interface{}) error
+}
+
+type serverDelegate struct {
+	*Server
+}
+
+// NewSinkManager builds a new SinkManager. It also creates ManagedSinks for
+// all EventSinks in the state store
+func NewSinkManager(ctx context.Context, delegate SinkDelegate, l hclog.Logger) *SinkManager {
+	m := &SinkManager{
+		ctx:                ctx,
+		delegate:           delegate,
+		updateSinkInterval: 30 * time.Second,
+		sinkSubscriptions:  make(map[string]*ManagedSink),
+		L:                  l.Named("sink_manager"),
+	}
+
+	return m
+}
+
+// EstablishManagedSinks creates and sets ManagedSinks for the Manager
+func (m *SinkManager) EstablishManagedSinks() error {
+	m.eventSinksWs = m.delegate.State().NewWatchSet()
+
+	iter, err := m.delegate.State().EventSinks(m.eventSinksWs)
+	if err != nil {
+		return err
+	}
+
+	var sinkIDs []string
+	for {
+		raw := iter.Next()
+		if raw == nil {
+			break
+		}
+		sink := raw.(*structs.EventSink)
+		sinkIDs = append(sinkIDs, sink.ID)
+	}
+
+	for _, id := range sinkIDs {
+		mSink, err := NewManagedSink(m.ctx, id, m.delegate.State, m.L)
+		if err != nil {
+			return fmt.Errorf("creating managed sink: %w", err)
+		}
+		m.sinkSubscriptions[id] = mSink
+	}
+	return nil
+}
+
+// Running specifies if the manager is currently running
+func (m *SinkManager) Running() bool {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	return m.running
+}
+
+// Stop stops the manager
+func (m *SinkManager) Stop() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if !m.running {
+		return
+	}
+	close(m.shutdownCh)
+	m.running = false
+}
+
+// Run is a long running function that starts all of the ManagedSinks.
+func (m *SinkManager) Run() error {
+	m.mu.Lock()
+	m.running = true
+	m.shutdownCh = make(chan struct{})
+	m.mu.Unlock()
+
+	errCh := make(chan SinkError)
+
+	updateSinks := time.NewTicker(m.updateSinkInterval)
+
+	run := func(id string, ms *ManagedSink) {
+		err := ms.Run()
+		select {
+		case <-m.ctx.Done():
+		case errCh <- SinkError{ID: id, Error: err}:
+		}
+	}
+
+START:
+	for id, ms := range m.sinkSubscriptions {
+		sid, sinkSub := id, ms
+		if !sinkSub.Running() {
+			go run(sid, sinkSub)
+		}
+	}
+
+	for {
+		select {
+		case <-m.ctx.Done():
+			return m.ctx.Err()
+		case <-m.shutdownCh:
+			return nil
+		case <-updateSinks.C:
+			if err := m.updateSinkProgress(); err != nil {
+				m.L.Warn("unable to update sink progress", "error", err)
+			}
+		case err := <-m.NewSinkWs().WatchCh(m.ctx):
```

Comment:
This is 👍 . I might use this sort of construct to implement https://github.com/hashicorp/nomad/issues/9058